hardware/battery/battery_management_system_protocol.ksy

```diff
@@ -0,0 +1,223 @@
+meta:
+  id: battery_management_system_protocol
+  title: Communication protocol of smart battery management systems from LLT power
+  license: CC0-1.0
+  ks-version: 0.8
+  endian: be
+
+doc: |
+  Many modern general purpose BMS include a UART/Bluetooth based communication interface.
+  After sending read requests they respond with various information's about the battery state in
+  a custom binary format.
+
+doc-ref: https://www.lithiumbatterypcb.com/Protocol%20English%20Version.rar
+
+seq:
+  - id: magic_start
+    contents: [0xdd]
+  - id: cmd
+    type: u1
+
+  - id: body
+    type:
+      switch-on: cmd
+      cases:
+        0xa5: read_req
+        0x5a: write_req
+        _: response(cmd)
+  
+  - id: checksum
+    size: 2
```

Comment:
Please write into `doc` key what format/algorithm the `checksum` field uses and from which bytes it is calculated.

Also, it's a good practice to add a [parse instance](https://doc.kaitai.io/user_guide.html#_instances_data_beyond_the_sequence) called `checksum_input` for example that can be used to obtain a byte array which serves as the basis for checksum calculation. If someone wants to check or write the checksum, they can easily access it.

It can look like this:

```ksy
instances:
  checksum_input:
    pos: 0
    size: 16 # or something like _sizeof - checksum._sizeof
```

Check https://github.com/kaitai-io/kaitai_struct/issues/84 for more information on the compile-time `sizeof` operator.

Reply:
Doc added in 66778d83ab549a2fda38d8b2c65b79dc25902c36

A `checksum_input` would indeed be very useful. Thanks for the suggestion. 
But I can't get it to work, in theory this is just the raw bytes of `body`, I tried:
```ksy
  checksum_input:
    pos: 2
    size: body._sizeof
```
But this does not work with current implementation because `dynamic sized type`.

Should this not also work with substreams?
```ksy
  checksum_input:
    io: _root.body._io
    size-eos: true
```
Result: `/instances/checksum_input/io: don't know how to call method '_io' of object type 'KaitaiStructType'`
I also can't add a constant size there... sorry, I could not figure out how to do this.

Reply:
Is a `checksum_input` under these circumstances impossible in kaitai, or am I doing just something wrong here?

Reply:
> I tried:
> 
> ```yaml
>   checksum_input:
>     pos: 2
>     size: body._sizeof
> ```
> 
> But this does not work with current implementation because `dynamic sized type`.

Yes, the [`sizeof` operator](https://github.com/kaitai-io/kaitai_struct/issues/84) currently works only for what can be derived at compile time.

> Should this not also work with substreams?
> 
> ```yaml
>   checksum_input:
>     io: _root.body._io
>     size-eos: true
> ```
> 
> Result: `/instances/checksum_input/io: don't know how to call method '_io' of object type 'KaitaiStructType'`

Well, it turns out that referencing `_io` of a generic base type `KaitaiStruct` (which arose by combining different user-defined `types`) is not implemented. That's not intentional, [`_io` is a common method](https://doc.kaitai.io/user_guide.html#_user_defined_types_2) of all user-defined types, in fact it's possible to do it manually from application code even in static-typed languages like [Java](https://github.com/kaitai-io/kaitai_struct_java_runtime/blob/ea3abf89ee7745ac56c814202c4ddb26fc668e11/src/main/java/io/kaitai/struct/KaitaiStruct.java#L41), only in KSY compiler nobody has implemented it yet.

However, even if it is implemented, it will not help in this case, because `body` doesn't have its own substream (and thus `_root.body._io` would just return the same as `_root._io`). Substream is created if and only if there is some byte `size` delimitation, which can be done using keys `size`, `size-eos: true` or `terminator`. 

> Is a `checksum_input` under these circumstances impossible in kaitai, or am I doing just something wrong here?

I think the best thing you can do is to store [`_io.pos`](https://doc.kaitai.io/user_guide.html#_streams) before and after parsing `body` and then just positioning the `checksum_input` on those exact offsets:

```ksy
seq:
  - id: magic_start
    contents: [0xdd]
  - id: cmd
    type: u1
  - size: 0
    if: ofs_body_start < 0 # storing current position
  - id: body
    type:
      switch-on: cmd
      # ...
  - size: 0
    if: ofs_body_end < 0 # storing current position
  - id: checksum
    type: u2
    # ...
  # ...
instances:
  ofs_body_start:
    value: _io.pos
  ofs_body_end:
    value: _io.pos
  checksum_input:
    pos: ofs_body_start
    size: ofs_body_end - ofs_body_start
```

In fact, it's a direct workaround for https://github.com/kaitai-io/kaitai_struct/issues/84 not supporting run-time calculations.

Reply:
Thanks again! Added in 26f169a.